### PR TITLE
Completion should generate only start section when empty end section is found

### LINKED ISSUE #805: Completion should generate only start section when empty end section is found
To support correctly [surround with tag section](https://github.com/redhat-developer/vscode-quarkus/issues/581), we need generate only start section when empty end section is found.

Given this template

```html
{#}
```

completion shows for, etc, after for item apply, it generates:

```html
{#for item in items}
  {item.name}
{/for}
```

Given this template which contains an empty orphan end section:

```
{#}
{\}
```

completion shows for, etc, after for item apply, it SHOULD generate:

```
{#for item in items}
{\}
```

@JessicaJHee I'm working on this issue.

### PR DESCRIPTION
**Pull request by angelozerr**
Generate only start section when empty end section is foundCompletion should generate only start section when empty end section is found

Fixes #805

Signed-off-by: azerr <azerr@redhat.com>

### CODE REVIEW

**Comment by angelozerr** [comment]
@JessicaJHee here a new PR which do the samething than PR https://github.com/redhat-developer/quarkus-ls/pull/806 but without updating end tag section (since `{/}` is a valid end section). Here a demo:

SnippetCompletionOnStartTag

**Comment by angelozerr** [comment]
The code of this PT should be clean, you can review it.

**Comment by datho7561** [comment]
If you have an unnamed close tag right after what you're writing, then it will never generate a close tag for tag completion, even if there aren't enough unnamed close tags to match the content you are generating:

DoesntGenerateCloseTag

**Comment by angelozerr** [comment]
> If you have an unnamed close tag right after what you're writing, then it will never generate a close tag for tag completion, even if there aren't enough unnamed close tags to match the content you are generating:

I have improved template parser to take care of this usecase. Please retry it.

**Comment by angelozerr** [comment]
@JessicaJHee could you check if it is working with surround please.

**Comment by datho7561** [comment]
Given the following template:

```


|{#each items}

{#each items}
  {it.name}
{/each}

{#each items}

{#each items}
  {it.name}
{/each}



{/}|
```

If I select between the two `|`, then I get a lot of code action failure notifications

**Review by JessicaJHee — APPROVED** [review]
LGTM, works well with surround with command! Thank you for writing lots of tests :)

**Comment by angelozerr** [comment]
According [this comment](https://github.com/redhat-developer/quarkus-ls/pull/807#issuecomment-1452519160) it is not a good idea to change the behavior of the parser.

I revert my change and to fix the original issue  

```html
{#each items}
	{#|}
{/}
```

which didn't generate the full content of the snippet (only the first part of the section), I check if parent section of {# is closed. In this case, #each is not closed because the `{/}` is linked to `{#`. As parent section of `{#` (which is the #each) is not closed, the full content of the snippet must be generated.

@datho7561 please retry it and I suggest that you get https://github.com/redhat-developer/vscode-quarkus/pull/585 to play with surround too.

**Comment by angelozerr** [comment]
Here a demo:

SurroundSection

**Comment by datho7561** [comment]
Seems pretty good so far, I'm going to try it out with https://github.com/redhat-developer/vscode-quarkus/pull/585 and see how that goes...

**Review by datho7561 — CHANGES_REQUESTED** [review]
This works very well. I found a few spots where the code could be cleaned up a bit, but nothing major. Feel free to merge after you've addressed the comments.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java [hunk 1]
```diff
@@ -628,4 +638,20 @@ public static boolean isWhenSection(Section section) {
 		return sectionKind == SectionKind.SWITCH || sectionKind == SectionKind.WHEN;
 	}
 
-}
+	/**
+	 * Returns true if the given section has an empty end section {/] and false
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/parser/template/Section.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+import com.redhat.qute.ls.commons.snippets.DefaultSnippetContentProvider;
+import com.redhat.qute.ls.commons.snippets.ISuffixPositionProvider;
+import com.redhat.qute.ls.commons.snippets.Snippet;
+import com.redhat.qute.parser.template.Node;
+import com.redhat.qute.parser.template.Section;
+import com.redhat.qute.parser.template.Template;
+import com.redhat.qute.utils.QutePositionUtility;
+
+/**
+ * Snippet content provider customizer foe Qute sections used to generate..
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java]: I have updated Javadoc with a sample, hope you will understand more.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+			replaceRange = getReplaceRange(startExpr, endExpr, offset, template);
+		}
+	}
+
+	@Override
+	public Position findSuffixPosition(String suffix) {
+		if (startTagClosePosition != null) {
+			return startTagClosePosition;
+		}
+		return null;
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java]: Indeed :) Good catch.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+			hasParameters = !section.getParameters().isEmpty();
+
+			// Position of start section closed
+			if (!hasParameters && section.isStartTagClosed()) {
+				startTagClosePosition = template.positionAt(section.getStartTagCloseOffset() + 1);
+			}
+			generateOnlyFirstPartOfSection = section.hasEndTag();
+			if (generateOnlyFirstPartOfSection && section.hasEmptyEndTag()) {
+				// {#for}{/}
+				// Check if there is a parent section which is not cloded, in this case the {\}
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java]: fixed

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+	 * @param offset      the offset.
+	 * @return the limit start offset of the expression according to the current
+	 *         node.
+	 */
+	private static int getExprLimitStart(Node currentNode, int offset) {
+		if (currentNode == null) {
+			// should never occurs
+			return 0;
+		}
+		// if (currentNode.isText()) {
```
[the remark this answers is not in the document]
Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/completions/tags/QuteSnippetContentProvider.java]: fixed

**Review by datho7561 — APPROVED** [review]
Looks good. Thanks, Angelo!

**Pull request merged**